File: tests/TestScreenConfiguration.m
classdef TestScreenConfiguration < matlab.unittest.TestCase
    properties
        config
    end

    methods (TestMethodSetup)
        function createConfig(testCase)
            % Create a standard screen configuration for testing
            testCase.config = ScreenConfiguration(500, 300, 1920, 1080, 600);
        end
    end

    methods (Test)
        function testConstructor(testCase)
            % Verify properties are set correctly
            sc = ScreenConfiguration(400, 250, 1600, 900, 500);
            testCase.verifyEqual(sc.screen_size_x_mm, 400);
            testCase.verifyEqual(sc.screen_size_y_mm, 250);
            testCase.verifyEqual(sc.screen_res_x_pix, 1600);
            testCase.verifyEqual(sc.screen_res_y_pix, 900);
            testCase.verifyEqual(sc.viewing_distance_mm, 500);
        end

        function testPixToMm(testCase)
            [x_mm, y_mm] = testCase.config.pix_to_mm(960, 540);
            testCase.verifyEqual(x_mm, 250, 'AbsTol', 1e-10);
            testCase.verifyEqual(y_mm, 150, 'AbsTol', 1e-10);
        end

        function testMmToPix(testCase)
            [x_pix, y_pix] = testCase.config.mm_to_pix(250, 150);
            testCase.verifyEqual(x_pix, 960, 'AbsTol', 1e-10);
            testCase.verifyEqual(y_pix, 540, 'AbsTol', 1e-10);
        end

        function testMmToDeg(testCase)
            [azi, ele] = testCase.config.mm_to_deg(250, 0);
            testCase.verifyEqual(azi, atan2d(250, 600), 'AbsTol', 1e-10);
            testCase.verifyEqual(ele, 0, 'AbsTol', 1e-10);
        end

        function testDegToMm(testCase)
            azi = atan2d(250, 600);
            ele = 0;
            [x_mm, y_mm] = testCase.config.deg_to_mm(azi, ele);
            testCase.verifyEqual(x_mm, 250, 'AbsTol', 1e-10);
            testCase.verifyEqual(y_mm, 0, 'AbsTol', 1e-10);
        end

        function testPixToDeg(testCase)
            [azi, ele] = testCase.config.pix_to_deg(960, 540);
            expected_azi = atan2d(250, 600);
            expected_ele = atan2d(150, hypot(600, 250));
            testCase.verifyEqual(azi, expected_azi, 'AbsTol', 1e-10);
            testCase.verifyEqual(ele, expected_ele, 'AbsTol', 1e-10);
        end

        function testDegToPix(testCase)
            azi = atan2d(250, 600);
            ele = atan2d(150, hypot(600, 250));
            [x_pix, y_pix] = testCase.config.deg_to_pix(azi, ele);
            testCase.verifyEqual(x_pix, 960, 'AbsTol', 1e-10);
            testCase.verifyEqual(y_pix, 540, 'AbsTol', 1e-10);
        end

        function testScreenExtents(testCase)
            [x_deg, y_deg] = testCase.config.screen_extents();
            expected_x_deg = 2 * atan2d(250, 600);
            expected_y_deg = 2 * atan2d(150, 600);
            testCase.verifyEqual(x_deg, expected_x_deg, 'AbsTol', 1e-10);
            testCase.verifyEqual(y_deg, expected_y_deg, 'AbsTol', 1e-10);
        end
    end
end

File: r/ETDQualitizer/tests/testthat/test-screen_configuration.R
test_that("constructor sets fields correctly", {
  sc <- ScreenConfiguration$new(400, 250, 1600, 900, 500)
  expect_equal(sc$screen_size_x_mm, 400)
  expect_equal(sc$screen_size_y_mm, 250)
  expect_equal(sc$screen_res_x_pix, 1600)
  expect_equal(sc$screen_res_y_pix, 900)
  expect_equal(sc$viewing_distance_mm, 500)
})

test_that("pix_to_mm conversion is correct", {
  config <- ScreenConfiguration$new(500, 300, 1920, 1080, 600)
  mm <- config$pix_to_mm(960, 540)
  expect_equal(mm$x, 250, tolerance = 1e-8)
  expect_equal(mm$y, 150, tolerance = 1e-8)
})

test_that("mm_to_pix conversion is correct", {
  config <- ScreenConfiguration$new(500, 300, 1920, 1080, 600)
  pix <- config$mm_to_pix(250, 150)
  expect_equal(pix$x, 960, tolerance = 1e-8)
  expect_equal(pix$y, 540, tolerance = 1e-8)
})

test_that("mm_to_deg conversion is correct", {
  config <- ScreenConfiguration$new(500, 300, 1920, 1080, 600)
  deg <- config$mm_to_deg(250, 0)
  expect_equal(deg$azi, atan2(250, 600) * 180 / pi, tolerance = 1e-8)
  expect_equal(deg$ele, 0, tolerance = 1e-8)
})

test_that("deg_to_mm conversion is correct", {
  config <- ScreenConfiguration$new(500, 300, 1920, 1080, 600)
  azi <- atan2(250, 600) * 180 / pi
  ele <- 0
  mm <- config$deg_to_mm(azi, ele)
  expect_equal(mm$x, 250, tolerance = 1e-8)
  expect_equal(mm$y, 0, tolerance = 1e-8)
})

test_that("pix_to_deg returns correct azimuth and elevation", {
  config <- ScreenConfiguration$new(500, 300, 1920, 1080, 600)
  deg <- config$pix_to_deg(960, 540)
  expected_azi <- atan2(250, 600) * 180 / pi
  expected_ele <- atan2(150, sqrt(600^2 + 250^2)) * 180 / pi
  expect_equal(deg$azi, expected_azi, tolerance = 1e-10)
  expect_equal(deg$ele, expected_ele, tolerance = 1e-10)
})

test_that("deg_to_pix conversion is correct", {
  config <- ScreenConfiguration$new(500, 300, 1920, 1080, 600)
  azi <- atan2(250, 600) * 180 / pi
  ele <- atan2(150, sqrt(600^2 + 250^2)) * 180 / pi
  pix <- config$deg_to_pix(azi, ele)
  expect_equal(pix$x, 960, tolerance = 1e-1)
  expect_equal(pix$y, 540, tolerance = 1e-1)
})

test_that("pix_to_mm and mm_to_pix are consistent", {
  config <- ScreenConfiguration$new(500, 300, 1920, 1080, 600)
  mm <- config$pix_to_mm(960, 540)
  pix <- config$mm_to_pix(mm$x, mm$y)
  expect_equal(pix$x, 960, tolerance = 1e-10)
  expect_equal(pix$y, 540, tolerance = 1e-10)
})
test_that("mm_to_deg and deg_to_mm are consistent", {
  config <- ScreenConfiguration$new(500, 300, 1920, 1080, 600)
  deg <- config$mm_to_deg(250, 0)
  mm <- config$deg_to_mm(deg$azi, deg$ele)
  expect_equal(mm$x, 250, tolerance = 1e-10)
  expect_equal(mm$y,   0, tolerance = 1e-10)
})

test_that("screen_extents returns correct values", {
  config <- ScreenConfiguration$new(500, 300, 1920, 1080, 600)
  extents <- config$screen_extents()
  expected_x_deg <- 2 * atan2(250, 600) * 180 / pi
  expected_y_deg <- 2 * atan2(150, 600) * 180 / pi
  expect_equal(extents$width, expected_x_deg, tolerance = 1e-8)
  expect_equal(extents$height, expected_y_deg, tolerance = 1e-8)
})
